lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -360,6 +361,100 @@ class TTNNOperandsWorkaroundsRewriter
+      }
+
+      ArrayAttr reshapedInputShapeAttr =
+          rewriter.getI32ArrayAttr(std::vector<int32_t>(
+              reshapedInputShape.begin(), reshapedInputShape.end()));
+
+      auto reshapedInputType =
+          RankedTensorType::Builder(inputType).setShape(reshapedInputShape);
+
+      ttnn::ReshapeOp preReshapeOp = rewriter.create<ttnn::ReshapeOp>(
+          loc, Type(reshapedInputType), op.getInput(), reshapedInputShapeAttr);
+
+      scatter_dim = scatter_dim + (4 - inputTypeShape.size());
+
+      reshapedInputShape[scatter_dim] =
+          static_cast<int32_t>(reshapedInputShape[scatter_dim] / scatter_num);
+
+      auto scatteredInputType =
+          RankedTensorType::Builder(inputType).setShape(reshapedInputShape);
+
+      ttnn::ReduceScatterOp reduceScatterOp =
+          rewriter.create<ttnn::ReduceScatterOp>(
+              loc, Type(scatteredInputType), preReshapeOp.getResult(), device,
+              scatter_dim, op.getMathOp());
+
+      RankedTensorType outputType = mlir::cast<RankedTensorType>(op.getType());
+      SmallVector<int64_t> outputTypeShape(outputType.getShape());
+
+      std::vector<int64_t> reshapedOutputShape(4, 1);
+      for (size_t i = 0; i < outputTypeShape.size(); ++i) {
+        reshapedOutputShape[i + outputTypeShape.size()] = outputTypeShape[i];
+      }
+
+      auto reshapedOutputType =
+          RankedTensorType::Builder(outputType).setShape(reshapedOutputShape);
+
+      ttnn::AllGatherOp allGatherOp = rewriter.create<ttnn::AllGatherOp>(
+          loc, Type(reshapedOutputType), reduceScatterOp.getResult(), device,
+          scatter_dim);
+
+      ArrayAttr reshapedOutputShapeAttr = rewriter.getI32ArrayAttr(
+          std::vector<int32_t>(outputTypeShape.begin(), outputTypeShape.end()));
+
+      rewriter.replaceOpWithNewOp<ttnn::ReshapeOp>(op, Type(outputType),
+                                                   allGatherOp.getResult(),
+                                                   reshapedOutputShapeAttr);
+    } else {
+      // TODO(wooseoklee): Once all_reduce support in ttnn is stable, we can
+      // convert directly to ttnn.all_reduce.
+      inputTypeShape[scatter_dim] = inputTypeShape[scatter_dim] / scatter_num;
+      auto scatteredInputType =
+          RankedTensorType::Builder(inputType).setShape(inputTypeShape);
+
+      ttnn::ReduceScatterOp reduceScatterOp =
+          rewriter.create<ttnn::ReduceScatterOp>(loc, Type(scatteredInputType),
+                                                 op.getInput(), device,
+                                                 scatter_dim, op.getMathOp());
+
+      rewriter.replaceOpWithNewOp<ttnn::AllGatherOp>(
+          op, op.getType(), reduceScatterOp.getResult(), device, scatter_dim);
+    }
+    return success();
+  }
+};
+
 // Pass to apply workarounds to the operands of TTNN operations.
 class TTNNWorkarounds : public impl::TTNNWorkaroundsBase<TTNNWorkarounds> {
```

Comment:
By default TTNNWorkarounds pass is disabled. To unblock yourself, you can split down pipeline option workaroundPassEnabled into two separate pass configs:
layouotWorkaroundsEnabled and decompositionWorkaroundsEnabled

Additionally, we just need to change that the pipeline will always run this pass, but based on these two options we will execute specific pass rewrites.

The decomposition workaround option can be on by default, and I will enable the one for layout workarounds.